File: Visualizer/ShaCalc/Sha256Net/Block256.cs
using System;
using System.Collections.Generic;
using ShaCalc.Model;

namespace ShaCalc.Sha256Net
{
    class Block256
    {
        protected IntValue[] Ints = new IntValue[8];

        public IntValue this[int index]
        {
            get { return Ints[index]; }
        }

        public Block256(params IntValue[] ints)
        {
            for (int i = 0; i < ints.Length; i++)
                Ints[i] = ints[i];
        }

        public Block256(Block256 other)
            : this(other.Ints)
        {
        }

        public BitValue[] GetOutBits()
        {
            BitValue[] bits = new BitValue[Ints.Length * 32];
            for (int i = 0; i < Ints.Length; i++)
            {
                var iv = Ints[i];
                for (int n = 0; n < 4; n++)
                    for (int b = 0; b < 8; b++)
                    {
                        bits[i * 32 + n * 8 + b] = iv[(3 - n) * 8 + b];
                    }
            }
            return bits;
        }

        public byte[] ToBytes()
        {
            byte[] value = new byte[Ints.Length * 4];
            for (int i = 0; i < Ints.Length; i++)
            {
                uint v = Ints[i].Get();
                value[i * 4 + 3] = (byte)(v);
                value[i * 4 + 2] = (byte)(v >> 8);
                value[i * 4 + 1] = (byte)(v >> 16);
                value[i * 4] = (byte)(v >> 24);
            }
            return value;
        }
    }
}

File: Graphula/Graphula/block256.cpp
#include "block256.h"

Block256::Block256()
{
}

Block256::Block256(const IntValue &v0, const IntValue &v1, const IntValue &v2, const IntValue &v3, const IntValue &v4, const IntValue &v5, const IntValue &v6, const IntValue &v7)
{
    Ints[0] = v0;
    Ints[1] = v1;
    Ints[2] = v2;
    Ints[3] = v3;
    Ints[4] = v4;
    Ints[5] = v5;
    Ints[6] = v6;
    Ints[7] = v7;
}
